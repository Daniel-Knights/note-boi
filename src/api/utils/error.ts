import { AppError, ERROR_CODE, ErrorConfig } from '../../classes';

/** Catches hanging requests (e.g. due to server error). */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function throwFetchError<T extends (...args: any) => Promise<void>>(
  errorConfig: Omit<ErrorConfig<T>, 'message'>,
  originalError: unknown
): void {
  throw new AppError({
    ...errorConfig,
    message: 'Request failed',
    originalError,
  });
}

/** Catches note encryption errors. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function throwEncryptorError<T extends (...args: any) => Promise<void>>(
  errorConfig: Omit<ErrorConfig<T>, 'code' | 'message'>,
  originalError: unknown
): void {
  throw new AppError({
    ...errorConfig,
    code: ERROR_CODE.ENCRYPTOR,
    message: 'Note encryption/decryption failed',
    originalError,
  });
}
