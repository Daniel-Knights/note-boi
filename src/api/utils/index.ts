export * from './decorators';
export * from './error';
export * from './http';
