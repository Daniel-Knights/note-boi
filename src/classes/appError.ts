export const ERROR_CODE = {
  NONE: 0,
  UNKNOWN: 1,
  LOGIN: 2,
  SIGNUP: 3,
  LOGOUT: 4,
  // PUSH was 5
  // PULL was 6
  DELETE_ACCOUNT: 7,
  CHANGE_PASSWORD: 8,
  ENCRYPTOR: 9,
  FORM_VALIDATION: 10,
  SYNC: 11,
} as const;

type ErrorCodes = typeof ERROR_CODE;
type ErrorKey = keyof ErrorCodes;
type ErrorCode = ErrorCodes[ErrorKey];

// Using `unknown` here breaks inference
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ErrorConfig<T extends (...args: any) => void> = {
  code: ErrorCode;
  message?: string;
  originalError?: unknown;
  /** Function to call on 'Try again', and its arguments */
  retry?: {
    fn: T;
    args?: Parameters<T>;
  };
  /** Whether to display the error in a form and/or `SyncStatus` */
  display?: {
    form?: boolean;
    sync?: boolean;
  };
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class AppError<T extends (...args: any) => void = (...args: any) => void> {
  readonly code;
  readonly message;
  readonly display;
  readonly retryConfig;
  readonly originalError;

  constructor(config: ErrorConfig<T> = { code: ERROR_CODE.NONE }) {
    this.code = config.code;
    this.message = config.message;
    this.display = Object.freeze(config.display);
    this.retryConfig = Object.freeze(config.retry);
    this.originalError = config.originalError;

    this.retry = this.retry.bind(this);
  }

  get isNone() {
    return this.code === ERROR_CODE.NONE;
  }

  retry() {
    if (!this.retryConfig) return;

    const { fn, args } = this.retryConfig;

    return args ? fn(args) : fn();
  }
}
