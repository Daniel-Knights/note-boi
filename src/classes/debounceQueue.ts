/**
 * Queues and debounces async function calls.
 * If a function is cancelled after it has already begun executing, `isCancelled` can
 * be used to check if it should stop.
 */
export class DebounceQueue {
  #cancelled = new Map<number, { isRunning: boolean }>();
  #current: { id: number | null; isRunning: boolean } = {
    id: null,
    isRunning: false,
  };

  /**
   * Adds new task to the debounce queue.
   * If there is an existing task, it will be cleared.
   */
  add(cb: (isCancelled: () => boolean) => Promise<void>, delay?: number) {
    this.clear();

    const timeoutId = window.setTimeout(() => {
      if (timeoutId !== this.#current.id) return;

      this.#current.isRunning = true;

      cb(() => this.#isCancelled(timeoutId)).finally(() => {
        this.#cancelled.set(timeoutId, { isRunning: false });
      });
    }, delay);

    this.#current.id = timeoutId;
    this.#current.isRunning = false;
  }

  /**
   * Clears the current task in the debounce queue.
   * If the task is running, it will be marked as cancelled.
   */
  clear() {
    if (!this.#current.id) return;

    if (this.#current.isRunning) {
      this.#cancelled.set(this.#current.id, { isRunning: true });
    } else {
      window.clearTimeout(this.#current.id);

      this.#current.id = null;
    }
  }

  /**
   * Checks if task with the given timeout ID has been cancelled.
   */
  #isCancelled(timeoutId: number) {
    const cancelled = !!this.#cancelled.get(timeoutId);

    this.#cancelled.delete(timeoutId);

    return cancelled;
  }
}
