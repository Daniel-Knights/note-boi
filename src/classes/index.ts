export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './storage';
export * from './unsyncedNotesManager';
