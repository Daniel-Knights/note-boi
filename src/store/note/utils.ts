import { Note } from '../../classes';
import { isEmptyNote } from '../../utils';

import { deleteNote } from './operations';
import { noteState } from './state';

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(uuid?: string): number {
  return noteState.notes.findIndex((nt) => nt.uuid === uuid);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(uuid?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.uuid === uuid);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.uuid === noteState.selectedNote.uuid ||
    noteState.extraSelectedNotes.some((nt) => nt?.uuid === note.uuid)
  );
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
export function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.uuid);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.uuid);
  }
}
