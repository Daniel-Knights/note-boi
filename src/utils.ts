import { event, invoke } from '@tauri-apps/api';
import { EventCallback, UnlistenFn } from '@tauri-apps/api/event';

import { TauriCommand } from './constant';
import type { Note } from './store/note';

/** `process.env.NODE_ENV === 'development'`. */
export function isDev(): boolean {
  return process.env.NODE_ENV === 'development';
}

/** Formats Unix time to date-time. */
export function unixToDateTime(unixTime: number): string {
  return Intl.DateTimeFormat([], {
    dateStyle: 'medium',
    timeStyle: isDev() ? 'long' : 'short',
  }).format(unixTime);
}

/** Gets key value from `localStorage` and parses it as JSON. */
export function localStorageParse(key: string): ReturnType<typeof JSON.parse> {
  return JSON.parse(localStorage.getItem(key) || '{}');
}

/** Returns `true` if string consists of only whitespace characters or is empty. */
export function isWhitespaceOnly(text?: string): boolean {
  return text?.trim() === '';
}

/** Returns true if note content is empty or note is `undefined`. */
export function isEmptyNote(note?: Note): boolean {
  if (!note) return true;

  return isWhitespaceOnly(note.content.title) && isWhitespaceOnly(note.content.body);
}

/** Calls {@link event.emit}, with stronger typing for `id`. */
export function tauriEmit<T>(id: 'login' | 'logout', payload?: T): void {
  event.emit(id, payload);
}

/** Calls {@link event.listen}, with stronger typing for `id`. */
export function tauriListen(
  id:
    | 'reload'
    | 'new-note'
    | 'delete-note'
    | 'push-notes'
    | 'login'
    | 'logout'
    | 'signup'
    | `tauri://${string}`,
  cb: EventCallback<unknown>
): Promise<UnlistenFn> {
  return event.listen(id, cb);
}

/** Calls {@link invoke}, with stronger typing for `cmd`. */
export function tauriInvoke<T>(
  cmd: TauriCommand,
  args?: Parameters<typeof invoke>[1]
): Promise<T> {
  return invoke(cmd, args);
}
