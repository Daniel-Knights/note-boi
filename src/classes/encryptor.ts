import { Note } from '../store/note';

import { KeyStore } from './keyStore';

export type EncryptedNote = Omit<Note, 'content'> & {
  content: string;
};

const SALT_LENGTH = 16;
const IV_LENGTH = 12;

const enc = new TextEncoder();
const dec = new TextDecoder();

function buffToBase64(buff: Uint8Array): string {
  return window.btoa(buff.reduce((data, byte) => data + String.fromCharCode(byte), ''));
}

function base64ToBuff(b64: string): Uint8Array {
  return Uint8Array.from(window.atob(b64), (c) => c.charCodeAt(0));
}

export class Encryptor {
  static #generatePasswordKey(password: string) {
    return crypto.subtle.importKey('raw', enc.encode(password), 'PBKDF2', false, [
      'deriveKey',
    ]);
  }

  static async setPasswordKey(password: string): Promise<void> {
    const passwordKey = await this.#generatePasswordKey(password);

    await KeyStore.storeKey(passwordKey);
  }

  static #deriveKey(
    passwordKey: CryptoKey,
    salt: Uint8Array,
    keyUsage: Array<'encrypt' | 'decrypt'>
  ): Promise<CryptoKey> {
    return crypto.subtle.deriveKey(
      {
        name: 'PBKDF2',
        salt,
        iterations: 100000,
        hash: 'SHA-256',
      },
      passwordKey,
      { name: 'AES-GCM', length: 256 },
      false,
      keyUsage
    );
  }

  static async #encryptData(secretData: string, passwordKey: CryptoKey): Promise<string> {
    const salt = crypto.getRandomValues(new Uint8Array(SALT_LENGTH));
    const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH));
    const derivedKey = await this.#deriveKey(passwordKey, salt, ['encrypt']);
    const encryptedContent = await crypto.subtle.encrypt(
      {
        name: 'AES-GCM',
        iv,
      },
      derivedKey,
      enc.encode(secretData)
    );

    const encryptedContentArr = new Uint8Array(encryptedContent);
    const buff = new Uint8Array(SALT_LENGTH + IV_LENGTH + encryptedContentArr.byteLength);

    buff.set(salt, 0);
    buff.set(iv, SALT_LENGTH);
    buff.set(encryptedContentArr, SALT_LENGTH + IV_LENGTH);

    return buffToBase64(buff);
  }

  static async #decryptData(
    encryptedData: string,
    passwordKey: CryptoKey
  ): Promise<string> {
    const encryptedDataBuff = base64ToBuff(encryptedData);
    const salt = encryptedDataBuff.slice(0, SALT_LENGTH);
    const iv = encryptedDataBuff.slice(SALT_LENGTH, SALT_LENGTH + IV_LENGTH);
    const data = encryptedDataBuff.slice(SALT_LENGTH + IV_LENGTH);
    const derivedKey = await this.#deriveKey(passwordKey, salt, ['decrypt']);
    const decryptedContent = await crypto.subtle.decrypt(
      {
        name: 'AES-GCM',
        iv,
      },
      derivedKey,
      data
    );

    return dec.decode(decryptedContent);
  }

  static async encrypt(plaintext: string, password?: string): Promise<string> {
    if (password) {
      await this.setPasswordKey(password);
    }

    const passwordKey = await KeyStore.getKey();

    return this.#encryptData(plaintext, passwordKey);
  }

  static async decrypt(ciphertext: string, password?: string): Promise<string> {
    if (password) {
      await this.setPasswordKey(password);
    }

    const passwordKey = await KeyStore.getKey();

    return this.#decryptData(ciphertext, passwordKey);
  }

  static async encryptNotes(notes: Note[], password?: string): Promise<EncryptedNote[]> {
    if (password) {
      await this.setPasswordKey(password);
    }

    const passwordKey = await KeyStore.getKey();

    const encryptedNotePromises = notes.map(async (nt) => {
      const encryptedNoteContent = await this.#encryptData(
        JSON.stringify(nt.content),
        passwordKey
      );

      const encryptedNote: EncryptedNote = {
        ...nt,
        content: encryptedNoteContent,
      };

      return encryptedNote;
    });

    return Promise.all(encryptedNotePromises);
  }

  static async decryptNotes(
    notes: (EncryptedNote | Note)[],
    password?: string
  ): Promise<Note[]> {
    if (password) {
      await this.setPasswordKey(password);
    }

    const passwordKey = await KeyStore.getKey();

    const decryptedNotePromises = notes.map(async (nt) => {
      if (typeof nt.content !== 'string') {
        return nt as Note;
      }

      const decryptedNoteContent = await this.#decryptData(nt.content, passwordKey);

      const decryptedNote: Note = {
        ...nt,
        content: JSON.parse(decryptedNoteContent),
      };

      return decryptedNote;
    });

    return Promise.all(decryptedNotePromises);
  }
}
