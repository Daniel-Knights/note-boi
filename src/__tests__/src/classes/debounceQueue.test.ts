import { DebounceQueue } from '../../../classes';
import { wait } from '../../utils';

describe('DebounceQueue', () => {
  it('add', () => {
    const dq = new DebounceQueue();
    const cb = vi.fn(() => Promise.resolve());

    vi.useFakeTimers();

    dq.add(cb, 1000);
    dq.add(cb, 1000);

    const timeoutId = dq.add(cb, 1000);

    vi.advanceTimersByTime(1000);
    vi.useRealTimers();

    expect(cb).toHaveBeenCalledTimes(1);

    assert.isDefined(timeoutId);
  });

  it('clear', () => {
    const dq = new DebounceQueue();
    const cb = vi.fn(() => Promise.resolve());

    vi.useFakeTimers();

    dq.add(cb, 1000);
    dq.add(cb, 1000);
    dq.add(cb, 1000);

    dq.clear();

    vi.advanceTimersByTime(1000);
    vi.useRealTimers();

    expect(cb).not.toHaveBeenCalled();
  });

  it('isCancelled', () => {
    const dq = new DebounceQueue();
    const cb = vi.fn(() => wait(1000));

    assert.isUndefined(dq.isCancelled(undefined));

    vi.useFakeTimers();

    let timeoutId = dq.add(cb, 1000);

    assert.isFalse(dq.isCancelled(timeoutId));

    dq.clear();

    assert.isFalse(dq.isCancelled(timeoutId));

    vi.advanceTimersByTime(1000);

    expect(cb).not.toHaveBeenCalled();

    vi.useFakeTimers();

    dq.add(cb, 1000);

    timeoutId = dq.add(cb, 1000);

    vi.advanceTimersByTime(1000);

    dq.clear();

    assert.isTrue(dq.isCancelled(timeoutId));

    vi.advanceTimersByTime(1000);
    vi.useRealTimers();
  });
});
