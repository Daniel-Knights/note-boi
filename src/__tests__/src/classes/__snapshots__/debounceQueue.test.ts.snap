// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DebounceQueue > add > KeyStore.getKey() 1`] = `undefined`;

exports[`DebounceQueue > add > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`DebounceQueue > add > Storage.get('THEME') 1`] = `null`;

exports[`DebounceQueue > add > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`DebounceQueue > add > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`DebounceQueue > add > Storage.get('USERNAME') 1`] = `null`;

exports[`DebounceQueue > add > allCalls 1`] = `[]`;

exports[`DebounceQueue > add > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`DebounceQueue > add > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`DebounceQueue > add > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`DebounceQueue > add > openedPopup 1`] = `undefined`;

exports[`DebounceQueue > add > selectedTheme 1`] = `"System"`;

exports[`DebounceQueue > add > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`DebounceQueue > add > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`DebounceQueue > clear > KeyStore.getKey() 1`] = `undefined`;

exports[`DebounceQueue > clear > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`DebounceQueue > clear > Storage.get('THEME') 1`] = `null`;

exports[`DebounceQueue > clear > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`DebounceQueue > clear > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`DebounceQueue > clear > Storage.get('USERNAME') 1`] = `null`;

exports[`DebounceQueue > clear > allCalls 1`] = `[]`;

exports[`DebounceQueue > clear > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`DebounceQueue > clear > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`DebounceQueue > clear > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`DebounceQueue > clear > openedPopup 1`] = `undefined`;

exports[`DebounceQueue > clear > selectedTheme 1`] = `"System"`;

exports[`DebounceQueue > clear > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`DebounceQueue > clear > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
